app/webpacker/lib/hooks/useStoredReducer.js

```diff
@@ -0,0 +1,33 @@
+import { useReducer } from 'react';
+
+/**
+ * This functions like the useReducer hook, but it fetches the state stored in
+ * local storage, via the given key, on subsequent uses.
+ *
+ * Do NOT call this twice with the same key - updating one of such a pair
+ * will not update the other's state.
+ */
+export default function useStoredReducer(reducer, initialState, key) {
+  let storedState;
+  try {
+    storedState = JSON.parse(localStorage.getItem(key));
+  } catch {
+    storedState = null;
+  }
+
+  function augmentedReducer(state, action) {
+    const newState = reducer(state, action);
+    localStorage.setItem(key, JSON.stringify(newState));
+    return newState;
+  }
+
+  const [state, dispatch] = useReducer(augmentedReducer, initialState, (value) => {
+    if (storedState === null) {
+      localStorage.setItem(key, JSON.stringify(value));
+      return value;
+    }
+    return storedState;
```

Comment:
As far as I can tell, you should be able to move the whole initialization block for `storedState` into this callback, as the variable `storedState` is only ever used in here.